R/markdown-link.R

```diff
@@ -174,6 +185,100 @@ parse_link <- function(destination, contents, state) {
   }
 }
 
+resolve_link_package <- function(topic, me = NULL, pkgdir = NULL) {
+  me <- me %||% roxy_meta_get("current_package")
+  # this is  from the roxygen2 tests, should not happen on a real package
+  if (is.null(me) || is.na(me) || me == "") return(NA_character_)
+
+  # if it is in the current package, then no need for package name, right?
+  if (has_topic(topic, me)) return(NA_character_)
+
+  # try packages in depends, imports, suggests first, error on name clashes
+  pkgs <- local_pkg_deps(pkgdir)
+
+  pkg_has_topic <- pkgs[map_lgl(pkgs, has_topic, topic = topic)]
+  pkg_has_topic <- map_chr(pkg_has_topic, function(p) {
+    p0 <- find_reexport_source(topic, p)
+    if (is.na(p0)) p else p0
+  })
+  pkg_has_topic <- unique(pkg_has_topic)
+  base <- base_packages()
+  if (length(pkg_has_topic) == 0) {
+    # fall through to check base packages as well
+  } else if (length(pkg_has_topic) == 1) {
+    if (pkg_has_topic %in% base) {
+      return(NA_character_)
+    } else {
+      return(pkg_has_topic)
+    }
+  } else {
+    cli::cli_warn(c(
+      "Topic {.val {topic}} is available in multiple packages: {.pkg {pkg_has_topic}}.",
+      i = "Qualify topic explicitly with a package name when linking to it."
+    ))
+  }
+
+  # try base packages as well, take the first hit,
+  # there should not be any name clashes, anyway
+  for (bp in base) {
+    if (has_topic(topic, bp)) return(NA_character_)
+  }
+
+  cli::cli_warn(c(
+    "Could not resolve link to topic {.val {topic}} in the dependencies or base packages.",
+    "i" = paste(
+      "If you haven'r documented {.val {topic}} yet, or just changed its name, this is normal.",
```

Comment:
I had a vague memory that roxygen2 maintains some index of topics for the current package, but I can't find any sign of it in the code, so I'm confident that this is the best you can do

Reply:
Yeah, we don't have the topics yet here. This was basically my second idea above, i.e. wait until we have the topics and do the substitution after that. Probably possible, but I think the warnings are also ok.